mlrun/k8s_utils.py

```diff
@@ -228,3 +230,322 @@ def handle_invalid(message):
+        ),
+        affinity=affinity,
+    )
+
+    # Remove preemptible nodes constraint
+    node_selector = _prune_node_selector(
+        mlconfig.get_preemptible_node_selector(),
+        enriched_node_selector=node_selector,
+    )
+
+    # Use anti-affinity only if no tolerations configured
+    if not preemptible_tolerations:
+        affinity = _override_required_during_scheduling_ignored_during_execution(
+            kubernetes.client.V1NodeSelector(
+                node_selector_terms=generate_preemptible_nodes_anti_affinity_terms()
+            ),
+            affinity,
+        )
+
+    return node_selector, tolerations, affinity
+
+
+def _handle_constrain_mode(
+    node_selector, tolerations, affinity, preemptible_tolerations
+):
+    tolerations = _merge_tolerations(tolerations, preemptible_tolerations)
+
+    affinity = _override_required_during_scheduling_ignored_during_execution(
+        kubernetes.client.V1NodeSelector(
+            node_selector_terms=generate_preemptible_nodes_affinity_terms()
+        ),
+        affinity=affinity,
+    )
+
+    return node_selector, tolerations, affinity
+
+
+def _handle_allow_mode(node_selector, tolerations, affinity, preemptible_tolerations):
+    for op in [
+        mlrun.common.schemas.NodeSelectorOperator.node_selector_op_not_in.value,
+        mlrun.common.schemas.NodeSelectorOperator.node_selector_op_in.value,
+    ]:
+        affinity = _prune_affinity_node_selector_requirement(
+            generate_preemptible_node_selector_requirements(op),
+            affinity=affinity,
+        )
+
+    node_selector = _prune_node_selector(
+        mlconfig.get_preemptible_node_selector(),
+        enriched_node_selector=node_selector,
+    )
+
+    tolerations = _merge_tolerations(tolerations, preemptible_tolerations)
+    return node_selector, tolerations, affinity
+
+
+def _merge_tolerations(
+    existing: list[kubernetes.client.V1Toleration],
+    to_add: list[kubernetes.client.V1Toleration],
+) -> list[kubernetes.client.V1Toleration]:
+    merged = existing.copy()
+    for toleration in to_add:
+        if toleration not in merged:
+            merged.append(toleration)
+    return merged
+
+
+def _prune_node_selector(
+    node_selector: dict[str, str],
+    enriched_node_selector: dict[str, str],
+):
+    """
+    Prunes given node_selector key from function spec if their key and value are matching
+    :param node_selector: node selectors to prune
+    """
+    # both needs to exists to prune required node_selector from the spec node selector
+    if not node_selector or not enriched_node_selector:
+        return
+
+    mlrun.utils.logger.debug("Pruning node selectors", node_selector=node_selector)
+    for key, value in node_selector.items():
+        if value:
+            spec_value = enriched_node_selector.get(key)
+            if spec_value and spec_value == value:
+                enriched_node_selector.pop(key)
+
+    return enriched_node_selector
```

Comment:
```suggestion
    return {
        key: value
        for key, value in enriched_node_selector.items()
        if node_selector.get(key) != value
    }
```
Your code skips pruning if `node_selector[key]` is `None` (even if `enriched_node_selector[key]` is also `None` and matching) - not sure if that was intentional. If it is then my suggestion is:
```suggestion
    return {
        key: value
        for key, value in enriched_node_selector.items()
        if not node_selector.get(key) or node_selector.get(key) != value
    }
```
And you should add a comment explaining why. If not, then use the first suggestion.